interfaces/prompting/patterns/render.go

```diff
@@ -0,0 +1,376 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package patterns
+
+import (
+	"bytes"
+	"strings"
+)
+
+// RenderConfig is a configuration of a render node.
+type RenderConfig interface {
+	// NextEx modifies the configuration to the next state, if any remain.
+	// Returns the length in bytes of the new configuration, the number of
```

Comment:
total length? full length? anyway configuration should probably be called variant, otherwise we have two terms Variant and Configuration, the 2nd one is fairly abused so ambiguous